Fix sanitization without config

HifzaanDev opened the pull request:
## Describe the changes that are made
- Load secrets directly from secret.yaml files when config.yaml is missing
- Parse and resolve template placeholders like `{{string .secret.SecretAccessKey}}`
- Added functions:
  - applySecretDesanitization() - applies secret resolution to expected responses
  - loadSecretsFromYAML() - finds and loads all secret.yaml files
  - loadSecretFile() - reads individual secret files
  - resolveSecretTemplates() - replaces template placeholders with actual values
 -Tests now work with sanitized secrets even when config.yaml is deleted.

## Links & References
-

**Closes:** #2981 


### 🔗 Related PRs
- NA
### 🐞 Related Issues
#2981 
### 📄 Related Documents
- NA

## What type of PR is this? (check all applicable)

- [X] 🍕 Feature


## Added e2e test pipeline?

- [X] 🙅 no, because they aren't needed


## Added comments for hard-to-understand areas?
- [X] 👍 yes


## Added to documentation?

- [X] 🙅 no documentation needed

## Are there any sample code or steps to test the changes?
- [X] 👍 yes, mentioned below
1.record using metadata flag
2.create secrets.yaml using sanitize
 3. confirm our secret keys are now hidden
4.delete config yaml file
5.run test
6.if test passes meaning we fixed the error


## Self Review done?
- [X] ✅ yes


## Any relevant screenshots, recordings or logs?
BERFORE:
<img width="1908" height="536" alt="ss0" src="https://github.com/user-attachments/assets/50554d6d-5f29-4223-b729-3e9ad86904ed" />
AFTER:
<img width="1908" height="572" alt="ss1" src="https://github.com/user-attachments/assets/b61c5656-1909-4cbe-8685-ef7a8fa76fc4" />




## Additional checklist:
- [X] Have you read the [Contributing Guidelines on issues](https://keploy.io/docs/keploy-explained/contribution-guide/)?
- [X] Have you followed the [PR Semantics guide](https://github.com/keploy/keploy/wiki/PR-Semantics) for naming this PR?
- [X] Have you followed the [Branch Semantics guide](https://github.com/keploy/keploy/wiki/Branch-Semantics) for naming your branch?

HifzaanDev commented:
recheck

officialasishkumar commented:
The changes made in match.go is not correct. we shouldn't be writing this for every match.go. The logic should be centralized and not called in match.go or any integrations.

officialasishkumar commented:
could you attach a video of you running test that covers the following scenario? 

- config.yaml and secret.yaml both exists (and there is key: value in secrets)  
- secret.yaml exists with no config.yaml (both when there is no secrets sanitized and when secrets are present in secret.yaml) - there will be no secrets sanitized in basic apps with no authentication like a simple addition api that you can create
- only config.yaml exists

HifzaanDev commented:
> <img alt="image" width="735" height="692" src="https://private-user-images.githubusercontent.com/87874775/493013799-dc8bad08-5aae-47a2-88e1-acc8662fc3f0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3MDM5NzEsIm5iZiI6MTc1ODcwMzY3MSwicGF0aCI6Ii84Nzg3NDc3NS80OTMwMTM3OTktZGM4YmFkMDgtNWFhZS00N2EyLTg4ZTEtYWNjODY2MmZjM2YwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI0VDA4NDc1MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNiNGEyMjkxNzVmYTY3YzMzOWE5NGE2N2FiMjdmMDMwMTU0YTIwYzY0Y2RmODAyYWFlNzNmMDQ5MGY0ZGFjMWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.kXcTcaLeSRLi8DsC2JgBapJRbM81aDs7yEpliRLRhtw">



> <img alt="image" width="735" height="692" src="https://private-user-images.githubusercontent.com/87874775/493013799-dc8bad08-5aae-47a2-88e1-acc8662fc3f0.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTg3MDQ2NDQsIm5iZiI6MTc1ODcwNDM0NCwicGF0aCI6Ii84Nzg3NDc3NS80OTMwMTM3OTktZGM4YmFkMDgtNWFhZS00N2EyLTg4ZTEtYWNjODY2MmZjM2YwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTI0VDA4NTkwNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkyNjAzOTEzMTVhYjUxMzg4NzEzOTMzM2Y5NzQyOTkyOTM4YTM0NmU1OTZhZjlmYjIwNmE4NGNmYmY2MzJjNjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.dLA1JIiBRspUXhMA_qmE0IyBJRcasuOVlp88TMVULA8">

Thank you again for the detailed review. I have addressed all the comments and pushed the updated changes to the PR.

Here is a summary of the fixes:

### Review 1: `.gitignore` file should not be touched
**Action:** Reverted the changes to the `.gitignore` file and removed it from the PR.

### Review 2: Premature return when `config.yaml` exists
**Action:** Fixed. I removed the incorrect `return` statement. The code now correctly proceeds to load secrets from `secret.yaml` in all scenarios, ensuring consistent behavior.
<img width="1217" height="285" alt="image" src="https://github.com/user-attachments/assets/54dc2c8b-f7ef-4bc2-9c08-346f59e4c149" />


### Review 3: Add `filePath` to debug log
**Action:** Done. The debug log has been updated to include the `filePath` as requested.
<img width="1181" height="103" alt="image" src="https://github.com/user-attachments/assets/71475f22-e68c-4420-83cf-da43165ac696" />


Please let me know if there is anything else. Thank you!

HifzaanDev commented:
@officialasishkumar , 
### Test Scenarios for Secret Management

Sir, Here is a summary of the test flows that verify the fix is working correctly in all expected scenarios.

#### Scenario 1: `config.yaml` and `secret.yaml` both exist
-   **Setup:** Both `config.yaml` and a sanitized `secret.yaml` are present in the test set directory.
-   **Action:** Run the `keploy test` command.
-   **Expected Outcome:** ✅ **PASS**. The test succeeds as Keploy loads the configuration from `config.yaml` and then correctly uses `secret.yaml` to de-sanitize the test cases before comparison.

#### Scenario 2: Only `secret.yaml` exists (Core Fix Verification)
-   **Setup:** A sanitized `test.yaml` and its corresponding `secret.yaml` are present, but the `config.yaml` file has been deleted.
-   **Action:** Run the `keploy test` command.
-   **Expected Outcome:** ✅ **PASS**. This confirms the main bug is fixed. The code correctly creates a default configuration in memory and then proceeds to load and apply the secrets from `secret.yaml`, allowing the test to pass.

#### Scenario 3: Only `config.yaml` exists (Sanitized Test Case)
-   **Setup:** A sanitized `test.yaml` and `config.yaml` are present, but the `secret.yaml` file has been deleted.
-   **Action:** Run the `keploy test` command.
-   **Expected Outcome:** ❌ **FAIL**. This is correct behavior. The test fails because the test case contains placeholders, but the `secret.yaml` file is missing. Keploy correctly reports a diff between the sanitized and plain-text responses.

#### Scenario 4: Only `config.yaml` exists (Unsanitized Test Case)
-   **Setup:** A plain-text, unsanitized `test.yaml` (from a simple app with no secrets) and a `config.yaml` are present. No `secret.yaml` exists.
-   **Action:** Run the `keploy test` command.
-   **Expected Outcome:** ✅ **PASS**. This proves the fix does not interfere with normal test runs. The test passes because the plain-text test case correctly matches the plain-text live response.


**If all these test cases are what you expected i ll go ahead with creation of video and confirming it , if not please tell me whatever you find not to be correct.**

HifzaanDev commented:
> @HifzaanDev can you please provide a video of you running all the scenarios?


https://github.com/user-attachments/assets/0e149d6f-9dc4-49bf-a206-3dec0aef6b06

**Sir , this is the video of me testing all the possible edge cases**

officialasishkumar commented:
@HifzaanDev Added some review comments.

officialasishkumar commented:
Once my comments are addressed please ping me @HifzaanDev

officialasishkumar commented:
@HifzaanDev internal issue with the runners.. we will sort it out

officialasishkumar reviewed (commented):
Please remove your changes from all files except db.go since that is needed

HifzaanDev commented:
> @HifzaanDev Can you see it now?
yes sir , your major concerns about changing prompt.go , gen.go , and utils.go are valid , I made the changes to them because when i ran the golangci-lint run command these lines were shown in errors , even tho they were out of scope for this fix i still made the recommended changes by golangci-lint run, if you want i can delete those files from my pr

HifzaanDev commented on pkg/service/utgen/utils.go:
```diff
@@ -318,7 +318,7 @@ func createTestFile(testFilePath string, sourceFilePath string) (bool, error) {
 		}()
```
golangci-lint run shows this as potential error, i can revert

HifzaanDev commented on pkg/service/utgen/gen.go:
```diff
@@ -805,7 +805,7 @@ func (g *UnitTestGenerator) saveFailedTestCasesToFile() error {
 	}
 
-	_, err = fileHandle.WriteString(fmt.Sprintf("%s\n", builder.String()))
+	_, err = fmt.Fprintf(fileHandle, "%s\n", builder.String())
```
golangci-lint run shows this as potential error can be reverted

HifzaanDev replied:
> what's the reason for this change?

Using fmt.Fprintf is more idiomatic and efficient than WriteString(fmt.Sprintf(...)). It writes the formatted string directly to the file in one step, avoiding the creation of an unnecessary intermediate string in memory. It's a small but positive improvement for performance.

HifzaanDev commented on .gitignore:
```diff
@@ -1,65 +0,0 @@
-# Binaries for programs and plugins
-*.exe
-*.exe~
-*.dll
```
it doesnt show up in files changed section now its same as the original

HifzaanDev commented on pkg/platform/yaml/configdb/testset/db.go:
```diff
@@ -30,28 +30,46 @@ func (db *Db[T]) Read(ctx context.Context, testSetID string) (T, error) {
 
-	secretConfig.SetSecrets(secretValues)
+	// Set secrets into the config struct if supported
+	if secretConfig, ok := any(config).(models.Secret); ok && len(secretValues) > 0 {
```
Sir, the logic has been seperated into multiple lines as requested , let me know if you want any other changes regarding this.

officialasishkumar commented:
Why have you made changes in utgen2 package? 

Those are not relevant to this issue

officialasishkumar reviewed (approved):
LGTM!

officialasishkumar commented:
Hey,

The PR looks good

we are just having some internal changes on the workflows. Once they are done we can make this PR live.

officialasishkumar commented on pkg/platform/yaml/configdb/testset/db.go:
```diff
@@ -30,28 +30,55 @@ func (db *Db[T]) Read(ctx context.Context, testSetID string) (T, error) {
+		}
+	} else {
+		// Config file exists, unmarshal it
+		if err := yamlLib.Unmarshal(data, &config); err != nil {
```
can you put it like this 

```

err := yamlLib.Unmarshal(data, &config);
if err != nil {
```
